tooling/minimize/intrinsics/src/lib.rs

```diff
@@ -29,3 +33,109 @@ pub unsafe fn deallocate(ptr: *mut u8, size: usize, align: usize) {
     let layout = Layout::from_size_align(size, align).unwrap();
     unsafe { System.deallocate(ptr, layout); }
 }
+
+// This global keeps track of any join handles produced. It is needed
+// because the minirust intrinsic for spawn only returns an integer and
+// the join only takes an integer.
+static JOIN_HANDLES: Mutex<Vec<Option<JoinHandle<()>>>> = Mutex::new( Vec::new() );
+
+struct SendPtr<T>(*const T);
+unsafe impl<T> Send for SendPtr<T> {}
+
+pub fn spawn(fn_ptr: extern "C" fn(*const ()), data_ptr: *const ()) -> usize {
+    let mut join_handles = JOIN_HANDLES.lock().unwrap();
+
+    if join_handles.len() == 0 {
+        join_handles.push(None);
+    }
+
+    let ptr = SendPtr(data_ptr);
+    let handle = thread::spawn(
+        move || {
+            let ptr = ptr; // avoid per-field closure capturing
+            fn_ptr(ptr.0);
+        }
+    );
+    join_handles.push(Some(handle));
+
+    // Return the index of the element we just pushed.
+    join_handles.len()-1
+}
+
+pub fn join(thread_id: usize) {
+    let mut join_handles = JOIN_HANDLES.lock().unwrap();
+    let handle = join_handles[thread_id].take().unwrap();
+    handle.join().unwrap();
+}
+
+
+#[derive(PartialEq)]
+enum LockState {
+    Open,
+    Locked,
+}
+
+// We implement our own locks for the intrinsics. They do not have an implicit
+// logic to unlock and operate on integers. This is why they are implemented in this way.
```

Comment:
Operate on integers? Implicit logic to unlock? I am very confused.
```suggestion
// We cannot use the locks from the standard library since MiniRust locks are acquired
// and released with an integer ID, but the standard library requires an `&Mutex` for
// `acquire` and a `MutexGuard` for `release`.
```